File: src/lib/api/resources.ts
import { ApiClient } from "./client";
import {
  ResourceConfig,
  QuickAccessItem,
  QuickAccessItemInput,
  ResourceItem,
  ResourceItemInput,
} from "./types";

export class ResourcesApi extends ApiClient {
  // Public resource methods
  async getResourcesConfig(): Promise<{
    site: {
      title: string;
      description: string;
      domain: string;
    };
    title: string;
    description: string;
    footer: {
      title: string;
      description: string;
      contactEmail: string;
      buttonText: string;
    };
  }> {
    return this.request("/api/resources/config");
  }

  async getQuickAccess(): Promise<
    Array<{
      id: string;
      title: string;
      subtitle: string;
      phone: string;
      color: string;
      icon: string;
    }>
  > {
    return this.request("/api/resources/quick-access");
  }

  async getResourceItems(category?: string): Promise<
    Array<{
      id: number;
      title: string;
      url: string;
      description: string;
      category: string;
      phone?: string;
      address?: string;
      icon: string;
    }>
  > {
    const params = category ? `?category=${encodeURIComponent(category)}` : "";
    return this.request(`/api/resources/items${params}`);
  }

  async getResourceCategories(): Promise<string[]> {
    return this.request("/api/resources/categories");
  }

  async getResources(): Promise<{
    site: {
      title: string;
      description: string;
      domain: string;
    };
    title: string;
    description: string;
    quickAccess: Array<{
      id: string;
      title: string;
      subtitle: string;
      phone: string;
      color: string;
      icon: string;
    }>;
    resources: Array<{
      id: number;
      title: string;
      url: string;
      description: string;
      category: string;
      phone?: string;
      address?: string;
      icon: string;
    }>;
    footer: {
      title: string;
      description: string;
      contactEmail: string;
      buttonText: string;
    };
  }> {
    return this.request("/api/resources");
  }

  // Admin resource methods
  async adminGetResourceConfigs(): Promise<
    Array<{
      id: number;
      key: string;
      value: string;
      description?: string;
      created_date: string;
      updated_date: string;
    }>
  > {
    return this.request("/api/admin/resources/config");
  }

  async adminUpdateResourceConfig(
    id: number,
    data: { value?: string; description?: string }
  ): Promise<{ message: string; config: ResourceConfig }> {
    return this.request(`/api/admin/resources/config/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
    });
  }

  async adminCreateResourceConfig(data: {
    key: string;
    value: string;
    description?: string;
  }): Promise<{ message: string; config: ResourceConfig }> {
    return this.request("/api/admin/resources/config", {
      method: "POST",
      body: JSON.stringify(data),
    });
  }

  async adminGetQuickAccessItems(): Promise<
    Array<{
      id: string;
      title: string;
      subtitle: string;
      phone: string;
      color: string;
      icon: string;
    }>
  > {
    return this.request("/api/admin/resources/quick-access");
  }

  async adminGetQuickAccessItem(id: number): Promise<{
    id: string;
    title: string;
    subtitle: string;
    phone: string;
    color: string;
    icon: string;
  }> {
    return this.request(`/api/admin/resources/quick-access/${id}`);
  }

  async adminCreateQuickAccessItem(
    data: QuickAccessItemInput
  ): Promise<{ message: string; item: QuickAccessItem }> {
    return this.request("/api/admin/resources/quick-access", {
      method: "POST",
      body: JSON.stringify(data),
    });
  }

  async adminUpdateQuickAccessItem(
    id: number,
    data: Partial<QuickAccessItemInput>
  ): Promise<{ message: string; item: QuickAccessItem }> {
    return this.request(`/api/admin/resources/quick-access/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
    });
  }

  async adminDeleteQuickAccessItem(id: number): Promise<{ message: string }> {
    return this.request(`/api/admin/resources/quick-access/${id}`, {
      method: "DELETE",
    });
  }

  async adminGetResourceItems(): Promise<
    Array<{
      id: number;
      title: string;
      url: string;
      description: string;
      category: string;
      phone?: string;
      address?: string;
      icon: string;
    }>
  > {
    return this.request("/api/admin/resources/items");
  }

  async adminGetResourceItem(id: number): Promise<{
    id: number;
    title: string;
    url: string;
    description: string;
    category: string;
    phone?: string;
    address?: string;
    icon: string;
  }> {
    return this.request(`/api/admin/resources/items/${id}`);
  }

  async adminCreateResourceItem(
    data: ResourceItemInput
  ): Promise<{ message: string; item: ResourceItem }> {
    return this.request("/api/admin/resources/items", {
      method: "POST",
      body: JSON.stringify(data),
    });
  }

  async adminUpdateResourceItem(
    id: number,
    data: Partial<ResourceItemInput>
  ): Promise<{ message: string; item: ResourceItem }> {
    return this.request(`/api/admin/resources/items/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
    });
  }

  async adminDeleteResourceItem(id: number): Promise<{ message: string }> {
    return this.request(`/api/admin/resources/items/${id}`, {
      method: "DELETE",
    });
  }
}

File: backend/app/routes/resources.py
import json

from flask import Blueprint, current_app, jsonify, request

from app import db
from app.models.resource import QuickAccessItem, ResourceConfig, ResourceItem

bp = Blueprint("resources", __name__)


@bp.route("/api/resources/config", methods=["GET"])
def get_resources_config():
    """Get the resources page configuration including title, description, footer, and site info from database"""
    try:
        config = {}
        config_items = ResourceConfig.query.all()
        config_dict = {item.key: item.value for item in config_items}

        config["site"] = {
            "title": config_dict.get("site_title", "Community Website"),
            "description": config_dict.get(
                "site_description", "Helping connect people to the resources available to them."
            ),
            "domain": config_dict.get("site_domain", "community.local"),
        }

        config["title"] = config_dict.get("resources_title", "Local Resources")
        config["description"] = config_dict.get(
            "resources_description", "Essential links to local services and information"
        )

        footer_json = config_dict.get("resources_footer")
        if footer_json:
            try:
                config["footer"] = json.loads(footer_json)
            except json.JSONDecodeError:
                config["footer"] = None

        if not config.get("footer"):
            config["footer"] = {
                "title": config_dict.get("footer_title", "Missing a Resource?"),
                "description": config_dict.get(
                    "footer_description",
                    "If you know of an important local resource that should be included on this page, please let us know.",
                ),
                "contactEmail": config_dict.get("footer_contact_email", "contact@example.com"),
                "buttonText": config_dict.get("footer_button_text", "Suggest a Resource"),
            }

        return jsonify(config)
    except Exception as e:
        current_app.logger.error(f"Error getting resources config: {str(e)}")
        return jsonify({"error": "Failed to load resources configuration"}), 500


@bp.route("/api/resources/quick-access", methods=["GET"])
def get_quick_access():
    """Get quick access items for resources page"""
    try:
        items = (
            QuickAccessItem.query.filter_by(is_active=True)
            .order_by(QuickAccessItem.display_order, QuickAccessItem.id)
            .all()
        )
        return jsonify([item.to_dict() for item in items])
    except Exception as e:
        current_app.logger.error(f"Error getting quick access items: {str(e)}")
        return jsonify({"error": "Failed to load quick access items"}), 500


@bp.route("/api/resources/items", methods=["GET"])
def get_resource_items():
    """Get resource items, optionally filtered by category"""
    try:
        category = request.args.get("category")
        query = ResourceItem.query.filter_by(is_active=True)

        if category:
            query = query.filter_by(category=category)

        items = query.order_by(
            ResourceItem.category, ResourceItem.display_order, ResourceItem.title
        ).all()
        return jsonify([item.to_dict() for item in items])
    except Exception as e:
        current_app.logger.error(f"Error getting resource items: {str(e)}")
        return jsonify({"error": "Failed to load resource items"}), 500


@bp.route("/api/resources/categories", methods=["GET"])
def get_resource_categories():
    """Get unique categories from resource items"""
    try:
        result = (
            db.session.query(ResourceItem.category).filter(ResourceItem.is_active).distinct().all()
        )
        categories = [row[0] for row in result if row[0]]
        return jsonify(sorted(categories))
    except Exception as e:
        current_app.logger.error(f"Error getting resource categories: {str(e)}")
        return jsonify({"error": "Failed to load resource categories"}), 500


@bp.route("/api/site-config", methods=["GET"])
def get_site_config():
    """Get site-wide configuration from database"""
    try:
        config_items = ResourceConfig.query.all()
        config_dict = {item.key: item.value for item in config_items}

        response = jsonify(
            {
                "site": {
                    "title": config_dict.get("site_title", "Community Website"),
                    "description": config_dict.get(
                        "site_description",
                        "Helping connect people to the resources available to them.",
                    ),
                    "domain": config_dict.get("site_domain", "community.local"),
                    "tagline": config_dict.get("site_tagline", "Community Directory"),
                    "directoryDescription": config_dict.get(
                        "site_directory_description",
                        "Discover local resources and community information.",
                    ),
                    "copyright": config_dict.get("site_copyright", "2025"),
                    "copyrightHolder": config_dict.get("site_copyright_holder", "Community"),
                    "copyrightUrl": config_dict.get("site_copyright_url", "#"),
                    "shortName": config_dict.get("site_short_name", "Community"),
                    "fullName": config_dict.get("site_full_name", "Community Website"),
                },
                "pagination": {
                    "defaultLimit": int(config_dict.get("pagination_default_limit", "20")),
                },
            }
        )
        # Cache for 10 minutes (600 seconds)
        response.headers["Cache-Control"] = "public, max-age=600"
        return response
    except Exception as e:
        current_app.logger.error(f"Error getting site config: {str(e)}")
        return jsonify({"error": "Failed to load site configuration"}), 500


@bp.route("/api/resources", methods=["GET"])
def get_resources():
    """Get complete resources page data in format expected by frontend"""
    try:
        config_response = get_resources_config()
        if config_response.status_code != 200:
            raise Exception("Failed to get config")
        config = config_response.get_json()

        quick_access_response = get_quick_access()
        if quick_access_response.status_code != 200:
            raise Exception("Failed to get quick access")
        quick_access = quick_access_response.get_json()

        items_response = get_resource_items()
        if items_response.status_code != 200:
            raise Exception("Failed to get resource items")
        resource_items = items_response.get_json()

        result = {
            "site": config["site"],
            "title": config["title"],
            "description": config["description"],
            "quickAccess": quick_access,
            "resources": resource_items,
            "footer": config["footer"],
        }

        response = jsonify(result)
        # Cache for 5 minutes (300 seconds) - resources change less frequently
        response.headers["Cache-Control"] = "public, max-age=300"
        return response
    except Exception as e:
        current_app.logger.error(f"Error getting complete resources data: {str(e)}")
        return jsonify({"error": "Failed to load resources data"}), 500
